drivers/counter/counter_native_posix.c

```diff
@@ -11,33 +11,81 @@
 
 #define DT_COUNTER_LABEL counter0
 #define DRIVER_CONFIG_INFO_FLAGS (COUNTER_CONFIG_INFO_COUNT_UP)
-#define DRIVER_CONFIG_INFO_CHANNELS 1
+#define DRIVER_CONFIG_INFO_CHANNELS CONFIG_COUNTER_NATIVE_POSIX_CHANNELS_NUMBER
 #define COUNTER_NATIVE_POSIX_IRQ_FLAGS (0)
 #define COUNTER_NATIVE_POSIX_IRQ_PRIORITY (2)
 
 #define COUNTER_PERIOD (USEC_PER_SEC / CONFIG_COUNTER_NATIVE_POSIX_FREQUENCY)
 #define TOP_VALUE (UINT_MAX)
 
-static struct counter_alarm_cfg pending_alarm;
-static bool is_alarm_pending;
+struct counter_alarm {
+	struct counter_alarm_cfg config;
+	bool active;
+};
+
+static struct counter_top_cfg top_cfg;
+static struct counter_alarm alarms[DRIVER_CONFIG_INFO_CHANNELS];
 static const struct device *device;
 
+static bool any_alarm_active(void)
+{
+	for (int ch_id = 0; ch_id < DRIVER_CONFIG_INFO_CHANNELS; ch_id++) {
+		if (alarms[ch_id].active) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static void schedule_next_irq(void)
+{
+	uint32_t next_value = top_cfg.ticks;
+	uint32_t current_value = hw_counter_get_value();
+
+	for (int ch_id = 0; ch_id < DRIVER_CONFIG_INFO_CHANNELS; ch_id++) {
+		if (alarms[ch_id].active &&
+		    alarms[ch_id].config.ticks >= current_value &&
+		    alarms[ch_id].config.ticks < next_value) {
+			next_value = alarms[ch_id].config.ticks;
+		}
+	}
+
+	hw_counter_set_target(next_value);
```

Comment:
> i'm not familiar with native_posix interrupts but don't you need to protect against being interrupted when scheduling next interrupt?

@nordic-krch Do you mean the case of being interrupted while executing this code?

I think this won't be an issue since interrupt controller in native_posix is implemented as HW model. HW models code is executed only when cpu is sleeping or k_busy_wait was called.

https://docs.zephyrproject.org/latest/boards/posix/native_posix/doc/index.html#architecture

Reply:
then we are good. i was not aware of that.

Reply:
or when an interrupt is enabled